File: 0005_Longest_Palindromic_Substring/main.go
package main

import (
	"fmt"
	"strings"
)

//=======================================================================================================================

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func longestPalindrome(s string) string {
	tmpS := "^#" + strings.Join(strings.Split(s, ""), "#") + "#$"
	lenTmpS := len(tmpS)
	palindromeRadianList := make([]int, lenTmpS)
	left, right := 0, 0
	maxLen := 0
	centerIndex := 0

	for i := 1; i < lenTmpS-1; i++ {
		if right > i {
			palindromeRadianList[i] = min(right-i, palindromeRadianList[2*left-i])
		}

		for tmpS[i+1+palindromeRadianList[i]] == tmpS[i-1-palindromeRadianList[i]] {
			palindromeRadianList[i]++
		}

		if i+palindromeRadianList[i] > right {
			left, right = i, i+palindromeRadianList[i]
		}
	}
	for i, v := range palindromeRadianList {
		if v > maxLen {
			maxLen = v
			centerIndex = i
		}
	}
	return s[(centerIndex-maxLen)/2 : (centerIndex+maxLen)/2]
}

//=======================================================================================================================

func colorize(text string, colorCode string) string {
	return fmt.Sprintf("\033[%sm%s\033[0m", colorCode, text)
}

func printTestHeader() {
	header := "🚀 Testing longestPalindrome Implementation"
	line := "="
	for i := 0; i < len(header); i++ {
		line += "="
	}

	fmt.Printf("\n%s\n", line)
	fmt.Println(header)
	fmt.Printf("%s\n\n", line)
}

func printTestCase(s string, expected, result string, status string) {
	cyan := "36"  // ANSI cyan
	green := "32" // ANSI green
	red := "31"   // ANSI red

	fmt.Printf("%s s = %s\n",
		colorize("➡ Test Case:", cyan), s)
	fmt.Printf("   Expected: %s\n", expected)
	fmt.Printf("   Received: %s\n", result)

	color := green
	if status == "FAILED" {
		color = red
	}
	fmt.Printf("   Status: %s\n", colorize(status, color))
	fmt.Println("------------------------------------------------------------")
}

func testLongestPalindrome() {
	type testCase struct {
		s        string
		expected string
	}

	testCases := []testCase{
		// Basic cases
		{"babad", "bab"}, // "aba" is also valid
		{"cbbd", "bb"},
		{"a", "a"},
		{"ac", "a"},

		// Normal cases
		{"racecar", "racecar"},
		{"abacdfgdcaba", "aba"},
		{"abba", "abba"},
		{"abcba", "abcba"},

		// Edge cases
		{"", ""},
		{"aaaaaaaaaa", "aaaaaaaaaa"},
		{"abcdefghij", "a"},

		// Long strings
		{strings.Repeat("a", 1000), strings.Repeat("a", 1000)},
		{strings.Repeat("ab", 500), strings.Repeat("a", 1)},
		{"a" + strings.Repeat("b", 998) + "a", "a" + strings.Repeat("b", 998) + "a"},
		{strings.Repeat("abcba", 200), strings.Repeat("abcba", 200)},
		{strings.Repeat("abac", 250), "aba"},
		{"x" + strings.Repeat("y", 998) + "x", "x" + strings.Repeat("y", 998) + "x"},
		{strings.Repeat("abcdcba", 100), strings.Repeat("abcdcba", 100)},
		{strings.Repeat("a", 999) + "b", strings.Repeat("a", 999)},
		{strings.Repeat("abababab", 125), "abababa"}, // "bababab" is also valid
		{"z" + strings.Repeat("a", 999), strings.Repeat("a", 999)},
	}

	printTestHeader()

	passed := 0
	failed := 0

	for _, tc := range testCases {
		result := longestPalindrome(tc.s)

		var status string
		if isPalindrome(result) && (len(result) == len(tc.expected) ||
			(len(tc.expected) > 0 && len(result) >= len(tc.expected))) {
			status = "PASSED"
			passed++
		} else {
			status = "FAILED"
			failed++
		}

		printTestCase(tc.s, tc.expected, result, status)
	}

	total := passed + failed
	fmt.Printf("\n📊 Test Results Summary:\n")
	fmt.Printf("%s PASSED: %d/%d\n", colorize("✅", "32"), passed, total)
	fmt.Printf("%s FAILED: %d/%d\n", colorize("❌", "31"), failed, total)

	if failed == 0 {
		fmt.Printf("\n%s\n", colorize("🎉 All test cases passed successfully!", "32;1"))
	} else {
		fmt.Printf("\n%s\n", colorize("⚠ Some tests failed. Please review the output.", "33;1"))
	}
}

func isPalindrome(s string) bool {
	for i := 0; i < len(s)/2; i++ {
		if s[i] != s[len(s)-1-i] {
			return false
		}
	}
	return true
}

func main() {
	testLongestPalindrome()
}

File: 0005_Longest_Palindromic_Substring/main.py
import unittest

#=======================================================================================================================

class Solution:
    def longestPalindrome(self, s: str) -> str:
        tmp_s = '#'.join(f'^{s}$')
        len_tmp_s = len(tmp_s)
        palindrome_radian_list = [0] * len_tmp_s
        left = right = 0

        for i in range(1, len_tmp_s - 1):
            palindrome_radian_list[i] = (right > i) and min(right - i, palindrome_radian_list[2 * left - i])
            tmp = palindrome_radian_list[i]
            while tmp_s[i + 1 + palindrome_radian_list[i]] == tmp_s[i - 1 - palindrome_radian_list[i]]:
                palindrome_radian_list[i] += 1

            if i + palindrome_radian_list[i] > right:
                left, right = i, i + palindrome_radian_list[i]
        max_len, center_index = max((n, i) for i, n in enumerate(palindrome_radian_list))
        return s[(center_index - max_len) // 2: (center_index + max_len) // 2]

#=======================================================================================================================

def is_palindrome(s: str) -> bool:
    return s == s[::-1]


class TestLongestPalindrome(unittest.TestCase):
    def setUp(self):
        self.sol = Solution()

    def test_basic_cases(self):
        test_cases = [
            ("babad", ["bab", "aba"]),
            ("cbbd", ["bb"]),
            ("a", ["a"]),
            ("ac", ["a", "c"]),
        ]

        for s, expected in test_cases:
            with self.subTest(s=s):
                result = self.sol.longestPalindrome(s)
                self.assertTrue(is_palindrome(result))
                self.assertIn(result, expected)
                self.assertEqual(len(result), max(len(x) for x in expected))

    def test_normal_cases(self):
        test_cases = [
            ("racecar", ["racecar"]),
            ("abacdfgdcaba", ["aba"]),
            ("abba", ["abba"]),
            ("abcba", ["abcba"]),
        ]

        for s, expected in test_cases:
            with self.subTest(s=s):
                result = self.sol.longestPalindrome(s)
                self.assertTrue(is_palindrome(result))
                self.assertIn(result, expected)

    def test_edge_cases(self):
        test_cases = [
            ("aaaaaaaaaa", ["aaaaaaaaaa"]),
            ("abcdefghij", ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j"]),
        ]

        for s, expected in test_cases:
            with self.subTest(s=s):
                result = self.sol.longestPalindrome(s)
                self.assertTrue(is_palindrome(result))
                self.assertIn(result, expected)

    def test_long_strings(self):
        test_cases = [
            ("a" * 1000, ["a" * 1000]),
            ("a" + "b" * 998 + "a", ["a" + "b" * 998 + "a"]),
            ("abcba" * 200, ["abcba" * 200]),
            ("x" + "y" * 998 + "x", ["x" + "y" * 998 + "x"]),
            ("abcdcba" * 100, ["abcdcba" * 100]),
            ("a" * 999 + "b", ["a" * 999]),
            ("z" + "a" * 999, ["a" * 999]),
        ]

        for s, expected in test_cases:
            with self.subTest(s=s):
                result = self.sol.longestPalindrome(s)
                self.assertTrue(is_palindrome(result))
                self.assertIn(result, expected)

        for s, expected_len in test_cases:
            with self.subTest(s=s):
                result = self.sol.longestPalindrome(s)
                self.assertTrue(is_palindrome(result))


if __name__ == '__main__':
    unittest.main(verbosity=2)